### PR TITLE
ENH: Support export PyArray_API and PyUFunc_API from shared libraries

### PR DESCRIPTION
**Pull request by huangqinjin**
Currently, if multiple extension modules share same initialization, and the code is insde a **shared library**. e.g. [`boost::python::numpy::initialize()`](https://github.com/boostorg/python/blob/47d5bc76f69e20625214381c930a2fad5765e2b3/src/numpy/numpy.cpp#L26), we have to `import_array()` again for each module:
```cpp
boost::python::numpy::initialize();
import_array();
```

This PR adds two macros: `PY_ARRAY_ATTRIBUTE` and `PY_UFUNC_ATTRIBUTE`. To address the issue above, just add the following macro definitions to every compilation unit of every extension module:
```cpp
#define NO_IMPORT_ARRAY
#define PY_ARRAY_UNIQUE_SYMBOL the_same_array_symbol_as_in_initialization_library
#define PY_ARRAY_ATTRIBUTE __attribute__((visibility("default")))   // __declspec(dllexport) for Windows

#define NO_IMPORT_UFUNC
#define PY_UFUNC_UNIQUE_SYMBOL the_same_ufunc_symbol_as_in_initialization_library
#define PY_UFUNC_ATTRIBUTE __attribute__((visibility("default")))   // __declspec(dllexport) for Windows
```

### CODE REVIEW

**Comment by seberg** [comment]
Just to make sure I understand correctly.  You need to share the API table in a shared library, but are using restricted visibility for the API table symbol?

**Comment by huangqinjin** [comment]
Yes. The shared library is compiled with `-fvisibility=hidden`, so need extra attribute to export the API table.

**Comment by seberg** [comment]
I don't have a strong opinion on whether this is useful or the best approach, but it seems fine.

Could you look where `PY_ARRAY_UNIQUE_SYMBOL` is documented and add the documentation for `PY_ARRAY_ATTRIBUTE`?  I also wonder if we really need two macros and whether you can find a clearer name like `PY_ARRAY_API_TABLE_ATTRIBUTE`.

**Comment by huangqinjin** [comment]
The macro names are just follow the existing `PY_ARRAY_UNIQUE_SYMBOL`, as it is not named `PY_ARRAY_API_UNIQUE_SYMBOL`.

I have added the document, kindly help polish the language.

**Comment by charris** [comment]
This will need a release note.

File: doc/source/reference/c-api/array.rst [hunk 1]
```diff
@@ -3213,19 +3215,31 @@ the C-API is needed then some additional steps must be taken.
     header file as long as you make sure that NO_IMPORT_ARRAY is
     #defined before #including that file.
 
-    Internally, these #defines work as follows:
+    If the single extension module not only consists of multiple
+    compilation units, but also spans multiple binaries, instead of
+    invoking ``import_array`` in every binary, you can leverage the
+    ``PY_ARRAY_ATTRIBUTE`` macro. Internally, the C-API is actually
```
Comment (seberg) [inline doc/source/reference/c-api/array.rst:3221]: I feel this should be clear about the hidden visibility.  The details can be in the bullet points below.

**Review by seberg — COMMENTED** [review]
I would still much prefer to call it `API_ARRAY_ATTRIBUTE` or even API_TABLE.  The "unique symbol" is a bit underspecific name as well, but I don't feel just because it is, the new name needs to be also.

@rossbar would you have time to smoothen the docs here?  The docs could use a little bit of love, but I don't have a great rephrasing idea right now.

**Comment by seberg** [comment]
Thanks, I'll close this.  We eventually did this with a single attribute and also rewriting the docs.  Sorry that this stalled, but it is now available (in 2.1 I think, or will be in 2.2).

I will note that I very much **discourage** doing this.  It is a bad idea unless you bundle and control all of the shared libraries in questions in a single application.  Otherwise they might have been compiled against different NumPy versions.
And that means that you can get random crashes/undefined behavior: so if you do this, all compatibility/supports bets are off.

**Pull request closed without merging**